src/colmap/sensor/models.h

```diff
@@ -193,6 +213,35 @@ struct BaseCameraModel {
   static inline void IterativeUndistortion(const T* params, T* u, T* v);
 };
 
+// Base model for Fisheye camera models
+template <typename CameraModel>
+struct BaseFisheyeCameraModel : public BaseCameraModel<CameraModel> {
+  template <typename T>
+  static inline void FisheyeFromNormal(const T u, const T v, T* uu, T* vv) {
+    *uu = u;
+    *vv = v;
+    const T r = ceres::sqrt(u * u + v * v);
+    if (r > T(std::numeric_limits<double>::epsilon())) {
+      const T theta = ceres::atan(r);
+      *uu *= theta / r;
+      *vv *= theta / r;
+    }
+  }
+
+  template <typename T>
+  static inline void NormalFromFisheye(const T uu, const T vv, T* u, T* v) {
+    *u = uu;
+    *v = vv;
+    const T theta = ceres::sqrt(uu * uu + vv * vv);
+    const T theta_cos_theta = theta * ceres::cos(theta);
```

Comment:
Can we not handle the theta > PI/2 case here in the if-condition instead? theta_cos_theta would be zero in this case and then negative?

Reply:
It will be positive again when it becomes large enough. But yes we can do the check here on theta. 

However, how should we do it here? Since we don't have a boolean flag in this function to indicate whether it is viable. And since this is also used in Ceres context, I did not see a good way to integrate here.

Reply:
We could just clamp it to the range [EPS, PI - EPS] ?

Reply:
No. The problem is that for Fisheye cameras, there is only a range of pixel coordinates being valid. That is to say, there is only a constrained set of pixels that can be projected onto because of the Fisheye mapping. A distance of pi/2 to the image center indicates an infinity in the normal coordinate, and a distance larger than pi/2 can never be a valid pixel in the Fisheye cameras.

Reply:
Yes, I understand. But clamping to the suggested range above would solve that problem? Points outside that range would just be moved to the boundary of the valid pixels using this approach?

Reply:
Let's keep it as is. It seems like nobody hit this issue in practice before. Thanks.